client/webserver/site/src/js/forms.ts

```diff
@@ -1124,26 +1124,42 @@ export class WalletWaitForm {
       Doc.hide(page.syncSpinner)
       Doc.show(page.syncUncheck)
     }
-    page.progress.textContent = String(Math.round(prog * 100))
-
-    // The remaining time estimate must be based on more than one progress
-    // report. We'll cache up to the last 20 and look at the difference between
-    // the first and last to make the estimate.
+    page.progress.textContent = (prog * 100).toFixed(1)
+
+    if (prog >= 0.999) {
+      // The final stage of wallet sync process can take a while (it might hang
+      // at 99.9% for many minutes, indexing addresses for example), the simplest
+      // way to handle it is to keep displaying "finishing up" message until the
+      // sync is finished, since we can't reasonably show it progressing over time.
+      Doc.hide(page.syncRemaining)
+      Doc.show(page.syncFinishingUp)
+      Doc.show(page.syncRemainBox)
+      return
+    }
+    // Before we get to 99.9% the remaining time estimate must be based on more
+    // than one progress report. We'll cache up to the last 20 and look at the
+    // difference between the first and last to make the estimate.
     const cacheSize = 20
     const cache = this.progressCache
     cache.push({
       stamp: new Date().getTime(),
       progress: prog
     })
+    if (cache.length < 2) {
+      // Can't meaningfully estimate remaining until we have at least 2 data points.
+      return
+    }
     while (cache.length > cacheSize) cache.shift()
-    if (cache.length === 1) return
```

Comment:
This change is curious. What's the purpose?

Reply:
It's equivalent, from what I understand, but
- failing fast (before `while (cache.length > cacheSize) cache.shift()`)
- and having a doc-comment for what exactly the intent is

makes it more readable/explicit for me.